### ISSUE TITLE
Test user experience on Windows 10

### ISSUE THREAD

**Issue by samWson** [exercism/pharo-smalltalk#31]
The idea is to pretend you are a student starting the Pharo language track for the first time. This is the users perspective, **not** a dev.

- Follow the [installation instructions](https://github.com/exercism/pharo/blob/master/docs/INSTALLATION.md)
- Try out exercises
- Push exercise solutions
- Pull new exercises

And so on.

**Comment by gregcline** [comment]
Hey I'm not sure if this is helpful, but I'm a beginner interested in learning Pharo and I'm pretty excited about this track. If it would be helpful I could try going through this process since I am a student that would be trying the track for the first time.

**Comment by macta** [comment]
Hi @gregcline, we’re definitely interested in help. As you may have gathered we are getting close to launch but still sorting out a few details and also need to actually write more of the exercises.

Give us a few days and then it would be handy to see how fresh eyes react to the instructions- particularly as Pharo is a much different environment to those normally on Exercism (and hence why we’d like to get it live as more people should experience it and learn why it has such a passionate following).

The bit I need to unblock is how to actually download the exercises when in beta mode -currently exercism is showing an error.

**Comment by samWson** [comment]
@gregcline are you still interested in helping out with this issue? I'm looking at the Windows 10 experience right now and it would be great to have input from others.

Our current installation guide is focused on using the bash shell, which you can get on Windows 10 now, but I had some [problems](https://github.com/exercism/pharo/issues/88#issuecomment-414230260) using it. 

The best way to install Pharo for Windows will be direct from the website. If you haven't already done so then install Pharo for Windows from [here](http://pharo.org/web/download). Run the installer to install the Pharo launcher. Use the Pharo launcher to create a new image (Pharo 6.1) and start it up. Open up a playground window with `ctrl + o + w` and copy the following code snippet into it:
```smalltalk
Metacello new 
 baseline: 'Exercism'; 
 repository: 'github://exercism/pharo:master/dev/src';
 load.
```
Highlight all of the code, right click and select `Do it` or `ctrl + d`. Then paste `ExercismManager welcome.`  in the playground and `do it`. From there follow the [**Getting started**](https://exercism.io/tracks/pharo/installation) section of the **Install Pharo** guide. It should be the same across operating systems from there. Note my firewall blocked a couple of these steps, the same may happen to you. 

If you need more direct help we have a discord channel for this language track with the rest of the Pharo community [here](https://discordapp.com/channels/223421264751099906/475984272357064714). Feel free to ping me directly. You can also leave comments directly in this issue.

Let me know if you have any questions.

**Comment by gregcline** [comment]
So I just tried this, and perhaps I missed some instructions, but when I try to download the hello-world exercise I run into an error.
It first opens a command prompt window:
![pharo_cmd](https://user-images.githubusercontent.com/16009129/44370519-835a6300-a4a8-11e8-8813-4cc59f4bf8a0.PNG)
Where I'm not sure what to do. Then closing it triggers an exception saying:
![pharo_directory_err](https://user-images.githubusercontent.com/16009129/44370566-bd2b6980-a4a8-11e8-9160-cf7741d307c4.PNG)

I tried moving the Pharo image I created into the Exercism directory and importing the image into the launcher then running the image. I also tried moving it into a "Pharo" directory in the Exercism directory with the same result. I'm clearly just missing a step, but I'm not sure what it should be.

**Comment by macta** [comment]
This is the kind of in the wild testing we need.

First of all, is exercism the cli I stalked and running? Eg at the command line, if you type “exercism

**Comment by gregcline** [comment]
Yes on both counts. At least I think so. I definitely have the exercism client installed, I can fetch other track's exercises.

I've tried to put the image in an Exercism sub-directory in three ways.

1. I tried copying the folder containing the image into a folder called `pharo` within the base exercism folder and launched it by "importing from the file system" through the launcher. Same error.
2. I tried installing the launcher to that same `pharo` folder, then creating and launching an image entirely from the launcher. Same error.
3. I tried taking the contents of the "image" folder that the launcher creates and putting that in the `\Exercism\pharo` folder bare. Running this seems to just freeze the image when I close the command line window. I'm not totally sure what is happening here.

Something to note, if I try to download an exercise just using the exercism CLI by itself it tells me `Error: Solution not found`. I don't know if this could cause the freezing I encountered. Maybe I don't have access to the beta version of the track?

**Comment by macta** [comment]
Ah - there's a clue - and something that is related to us being in beta - you need to join the pharo track first. This is strictly until we are really live - which means you need to go to https://exercism.io/tracks/pharo and press join.

If we can get you going, we can go back and better test for this case - albeit, its a bit artificial (if this is what it is). I'm guessing you've then not seen the user installation guide if this is the case.

**Comment by gregcline** [comment]
Yeah, that's my bad. Normally a user wouldn't be trying to download exercises without being in the track. 

My image still seems to be freezing even after I've joined the track. I can download the exercise using just the CLI but something isn't set up right with my installation of Pharo. Is there somewhere specific I need to install the launcher?

**Comment by macta** [comment]
It’s not the launcher - it’s the image which has to be in the same directory as where exercism will install the examples.

I think launcher makes this too hard and it was why I used zero conf command line installs for Linux and OS X.

If we can’t figure it out, might need a config setting for where the exercism default directory is.

**Comment by macta** [comment]
Eg

./hello-world
./pharo.image
./Pharo.changes

**Comment by gregcline** [comment]
So is opening a command prompt when running the the Exercism commands from within Pharo expected? 

Right now I have the image in the /Exercism/pharo folder, and running any of the commands opens a command prompt within which nothing happens. When I close the prompt the image seems to just freeze.

**Comment by macta** [comment]
Yes we are running the exercism cli to retrieve and submit exercises but possibly this doesn’t work so well on windows or I’m doing something wrong (this is how Pharo launcher works - so it is possible).

It seems to work fine on OS X - will need to dig into this.

How handy Are you at Pharo - maybe we can try some standalone stuff to figure this out.

**Comment by macta** [comment]
@samWson did you get it running in windows - which version? @gregcline what version of windows are you running in? I know that @samWson is working on some better instructions for windows users - so trying to determine if we have a problem that will need us to rejig the whole thing (in the future we did want to use http to call exercism and remove calling the cli - but the cli seemed like a good quick fix)

**Comment by gregcline** [comment]
I'm using Windows 10 Home Version 1803, build 17134.228.

I'm not very experienced with Pharo, just the very basics of syntax, the playground, debugger etc.

**Comment by macta** [comment]
Hey thanks for riding this out with us. @samWson any thoughts - I think you got something working right?

**Comment by samWson** [comment]
@macta I'll try and get on this issue this weekend. Going to be another busy one.

**Comment by samWson** [comment]
@gregcline thanks for testing this out. I didn't go as far as you did the first time around, but when I followed our own instructions I seem to be getting all the same errors as you. I have joined the Pharo track in Exercism, and I'm on exactly the same version and build of Windows 10 as you.

@macta The error `'Pharo must be run from the default exercism directory!'` originates from `ExercismManager>>#verifyExercismResult:`. The error happens when you close the command prompt that opens after trying to fetch an exercise. This is actually the correct behavior I believe, because that method is looking for a file path from `osProcessResult`, and it is empty because nothing was entered at the command prompt.

The bug is probably more the command prompt opening when trying to fetch an exercise. I'm guessing it's in `ExercismManager>>#fetchFromExercismTo:`

```smalltalk
fetchFromExercismTo: package
	| result cmd exercise tonelLocator reader pkgName root exercisePkg |
	
	(exercise := UIManager default request: 'Enter a valid exercism exercise (e.g. hello-world):') ifNil: [ ^ self ].
	cmd := 'exercism download --track=pharo --exercise=' , exercise.
	
	UIManager default inform: 'Loading: ' , exercise.
	
	result := PipeableOSProcess waitForCommand: cmd.
	self verifyExercismResult: result.
	
	tonelLocator := FileLocator cwd / exercise.
	(reader := TonelReader on: tonelLocator fileName: tonelLocator pathString) loadDefinitions.
	reader definitions
		do: [ :def | 
			def load.
			def isOrganizationDefinition
				ifTrue: [ pkgName := def categories first ] ].
			
	root := RPackageOrganizer default packageNamed: 'Exercism'.
	exercisePkg := root packages detect: [ :p | p categoryName = pkgName ].
	
	SystemAnnouncer uniqueInstance announce: (RPackageRegistered to: exercisePkg).
	UIManager default inform: 'Success, Happy Coding'
```

The local variable `cmd` seems to be set to the correct command: `exercism download --track=pharo --exercise=hello-world`. 

On line 9 there is `result := PipeableOSProcess waitForCommand: cmd.`. This message send opens up the command prompt in Windows. I have a suspicion that this may not be operating system agnostic but I can't find any method or comment on the class that would confirm the theory.

@gregcline if your interested in learning a bit more about Pharo and our code, you have all the same code I just mentioned above in your image in the `ExercismTools` package. You can have a look through there and get an idea of whats happening.

@macta I hope that helps. If I get some time tomorrow I will try out the same testing on macOS or Linux and see what `PipeableOSProcess waitForCommand: cmd` does.

**Comment by macta** [comment]
Hi guys - I guess I’m going to have to find a windows machine ... on OS X the command prompt call is invisible - it calls the command, gets the result and continues automatically.

On Windows I’m wondering if the osprocess she’ll command works differently or if the command it’s trying to execute stalls on some way? 

I think I have understood that your images aren’t running from an exercism directory (see above?) - your in some Pharo launcher directory and not in /home/exercism/Pharo (homemis whatever you set the exercism home to be - see the install instructions in our readme.md for developers).

Can we confirm what happens if you run the cmd in the directory that Pharo is actually in - does it give an error and stall ?

**Comment by gregcline** [comment]
Sorry if I wasn't clear. I've tried both: image in `/home/exercism/Pharo` directory and image elsewhere. Both have the same result.

**Comment by macta** [comment]
Ok - guys can you try the following below in a playground (cmd G - or run triangle) - just to get a feel for how OSProcess works on windows (I'm on holiday so its tricky getting access to Windows - but a friend has promised to bring over a laptop).

```
cmd := 'ls -al *.image'. "osx"
cmd := 'dir *.image'. "win equivalent?"
result := PipeableOSProcess waitForCommand: cmd.
result outputAndError.
```

On OSX I get the following (no windows appears, it just gives a result)
```
"
#(
'-rw-r--r--@ 1 macta  staff  45016496 16 Aug 16:06 Pharo.image' 
''
)
"
```

If you do get a window that stays open, can you grab a screenshot of it it too? It should silently run is my understanding - however we always knew that shelling out to the command prompt was a bit of a hack.

**Comment by macta** [comment]
Also - I think I've understood that both of you have experience in other languages? @samWson - your a ruby guy right? I'm wondering if we can advance #32 by figuring out the http protocol for exercism and just not using OSProcess at all (this is actually the better way). If either of you can show me the magic in another language - we can easily port it to pharo too and use  `ZnEasy get:/ZnEasy post:data:` (and your welcome to use Pharo to figure it out too).

I'm balancing time on some other Pharo 7 stuff (and I'm supposed to be on holiday too) - so anything that can advance us would be helpful.

**Comment by macta** [comment]
I also had another thought - from the picture above of the command prompt - there are spaces in the Pharo subdirectory of Exercism - can we try renaming it to just `exercism/Pharo61` - grasping at straws but I know windows can be funny with paths and maybe we need to escape something better.

**Comment by gregcline** [comment]
So I was experimenting with `PipeableOSProcess` today and encountered exactly the issue I was having with the exercise fetch. I ran the code you suggested and also encountered it. It opens a shell as you can see here:
![using_pipeableos](https://user-images.githubusercontent.com/16009129/44623098-37d2fb00-a894-11e8-9c83-f33fe1ab58b1.PNG)

Then when I close the shell the Pharo UI is frozen.

Poking around in the exercism cli code I think the API communication stuff is generally in here: https://github.com/exercism/cli/tree/master/cmd downloading is in `download.go` and submitting is in `submit.go`. It looks like they build a simple GET with the track and exercise in the query params to get metadata. I think the struct at the bottom of `download` is the json payload coming back from the API. 

You might also have to look around in here: https://github.com/exercism/cli/tree/master/workspace to see how they're downloading the files. I don't have time to go through them right now but I'll take a look at them later.

**Comment by macta** [comment]
Hey - that’s a great start, I’ll try and look later.

If any of you get ahead of me - ZnEasy get: with a url is the simple wrapper for Pharo http comms (it calls lower level stuff , which we can do too).

You can experiment in a playground with urls and params and see what results come back.

**Comment by macta** [comment]
I also note there are some exercism cli updates that mention Windows paths as been a problem. 

Could you also try exercism update (but it seems that osprecess is flakey on Windows anyway)

**Comment by samWson** [comment]
Pharo launcher installed the image in a completely different place to exercisim, which for me is in `$HOME\AppData\Roaming\exercism`. That problem didn't cross my mind.

> can we try renaming it to just exercism/Pharo61

@macta I think this is a good idea, and I don't see much potential harm.

I upgraded the exercism cli to the latest, and it doesn't seem to make any difference. I'll poke around the cli with HTTP like you suggested.

**Comment by macta** [comment]
@samWson just to clarify, the rename to Pharo61 was for Windows only if using Pharo launcher - but I think we have established shelling into Windows is problematic (thinking about it - I think Pharo launcher itself uses a different mechanism). 

Well just dig into #31 - not worth wasting more effort on a hack anyway.

@gregcline bear with us, we’ll get a solution as I’m guessing Windows is a big platform on Exercism.

**Comment by macta** [comment]
@gregcline I just merged the first part to fixing this (with @samWson and @bencoman 's help).

This will let you download an exercise and try to solve it with our special TDD approach. I'm interested in how you get along with this. NOTE: i  think submitting will be broken when you try that (but lets try) as we haven't yet figured out the http exchange to do that part directly. But this should move us forward nicely.

**Comment by gregcline** [comment]
So I accidentally found a little improvement to note. Since I was used to the previous version I just breezed through the API token prompt and entered the exercise name there. Silly me. It resulted in a 401 (no surprise), but the actual exception was a key error on the returned JSON. Some custom error handling there would be nice.

Other than that, things went well! I was able to download the exercise and fumble my way through solving it. When I tried submitting, it still opened the command line and I tried to just manually run the command to submit through that window but got the following error:
![submitting](https://user-images.githubusercontent.com/16009129/44887672-6e958080-ac9b-11e8-9f81-e67de77567ab.PNG)

So no dice submitting, but closing the command prompt didn't freeze my Pharo image this time, so that's nice.

**Comment by macta** [comment]
@gregcline I've create a new PR for the token error handling (actually all of the error handling is a bit rough - but that one is an obvious problem as it also caches the token so I think you get a bad place with that and its worth a bit more attention).

I didn't expect the submit to work - i'm hoping that @samWson or @bencoman might do some more leg work to figure out the http for submission (I'm still a bit short on time to do that)

**Comment by samWson** [comment]
With #96 merged I tested a full cycle of an exercise in a new image on Windows 10: downloading an exercise -> writing a solution -> passing a test -> submitting the solution -> viewing it on Exercism.io. I didn't have any breaking problems.

@gregcline if your still interested and have the time, it would be great if you could test out the experience for us again. Any bugs, or pain points you find would be valuable to know.

**Comment by gregcline** [comment]
So following the Getting Started I'm currently getting a `This URL was not recognised` error when fetching an exercise. It looks like the URL being fetched is `/v1/solutions/latest/latest?track_id=pharo&exercise_id=hello-world`.
![wrong_url](https://user-images.githubusercontent.com/16009129/48662575-cd27fa80-ea51-11e8-8bf0-2ff4235e9a48.PNG)

Maybe it's a result of this code interaction?
![url_code](https://user-images.githubusercontent.com/16009129/48662579-d913bc80-ea51-11e8-9176-a7468ec4ff4d.PNG)

**Comment by gregcline** [comment]
I think changing line 4 of `ExercismCommand class>>#configureToken:` from `ApiPath := '/v1/solutions/latest` to just `ApiPath := '/v1/solutions'` fixes this. I tried changing `ExercismDownload >> #getSolutionData` but that caused a 404 on submit.

**Comment by bencoman** [comment]
@gregcline Yes are exactly right.  It should be that.  It looks like the merge of the "apikey-better-error-handling" branch reverted that [change required for submitting exercises](https://github.com/exercism/pharo/pull/110#issuecomment-431674368).
@samWson this will be the same for your issue #113.

Aside: something I find strange on github is that looking here... https://github.com/exercism/pharo/network 
the [6bdb5b commit on the 12 Nov](https://github.com/exercism/pharo/blob/6bdb5b018/dev/src/ExercismTools/ExercismCommand.class.st#L74) shows ```ApiPath := '/v1/solutions' ```
while the following [5aef1140 on the 15 Nov](https://github.com/exercism/pharo/blob/5aef114000/dev/src/ExercismTools/ExercismCommand.class.st#L58) shows ```    ApiPath := '/v1/solutions/latest' ```
but a [comparison of the two](https://github.com/exercism/pharo/compare/5aef114000...6bdb5b01) shows no changes.

**Comment by samWson** [comment]
@gregcline I tried out the latest master on Windows 10 and this problem seems to be fixed. You will have to start with a fresh image again but you shouldn't see this problem anymore.

**Comment by samWson** [comment]
I close this one issue now as #120 picks up testing for Pharo 7.
